### PR TITLE
version: make it possibe to get version info during runtime

### PR DESCRIPTION
**Pull request by kldeng**
make it possibe to get version info during runtime.

### CODE REVIEW

File: src/main.c [hunk 1]
```diff
@@ -86,7 +86,10 @@ int main(int argc, char *argv[])
 
     /* check if dpvs is running and remove zombie pidfile */
     if (dpvs_running(DPVS_PIDFILE)) {
-        fprintf(stderr, "dpvs is already running\n");
+        fprintf(stderr,
```
Comment (beacer) [inline src/main.c]: If the binary is changed for example upgraded, the version shown by this line is not the `DPVS` version which is actually running. So user may get confused.

Comment (kldeng) [reply src/main.c]: Not really get it. Shouldn't the binary be re-compiled and restarted when upgrades. So the user will get the version actually running. Please correct me if I am wrong. Thanks.

Comment (beacer) [reply src/main.c]: Assume the old version is 1.0, and it's already running. 
New image is 1.1, now try to execute 1.1, then it  will show
```
dpvs is already running
version 1.1, build on ....
```
User may think dpvs 1.1 is already running, but actually it's 1.0.

Comment (kldeng) [reply src/main.c]: yes, you're right. I've modified the PR by adding a application arguments parse before the EAL ones. Please review it, thanks!

File: src/main.c [hunk 2]
```diff
@@ -73,6 +74,62 @@ static int set_all_thread_affinity(void)
...
+
+    while((c = getopt_long(argc, argv, short_options, long_options, NULL)) != -1) {
+        switch(c) {
+            case 'v':
+                fprintf(stderr, "dpvs version: %s, build on %s\n", DPVS_VERSION, DPVS_BUILD_DATE);
+                exit(EXIT_SUCCESS);
+            case 'h':
+                dpvs_usage(prgname);
+                exit(EXIT_SUCCESS);
+            case '?':
```
Comment (beacer) [inline src/main.c]: getopt_long return the case `?` equals to `default`, or 'default' branch will never get hit.
```
case '?':
default:
    dpvs_usage(prgname);
    exit(...);
```

File: src/main.c [hunk 3]
```diff
@@ -73,6 +74,62 @@ static int set_all_thread_affinity(void)
...
+
+    struct option long_options[] = {
+        {"version", 0, NULL, 'v'},
+        {"help", 0, NULL, 'h'},
+        {NULL, 0, 0, 0}
+    };
+
+    optind = 1;
+
+    while((c = getopt_long(argc, argv, short_options, long_options, NULL)) != -1) {
```
Comment (beacer) [inline src/main.c]: Need space between `while` and `(`
```
while ((c = ...) ..) {
}
```

File: src/main.c [hunk 4]
```diff
@@ -73,6 +74,62 @@ static int set_all_thread_affinity(void)
...
+        switch(c) {
```
Comment (beacer) [inline src/main.c]: spaces,
```
switch (c) {
```

File: src/main.c [hunk 5]
```diff
@@ -73,6 +74,62 @@ static int set_all_thread_affinity(void)
...
+
+static int parse_app_args(int argc, char **argv)
+{
+    const char *short_options = "vh";
+    char *prgname = argv[0];
+    int c, ret = -1;
+
+    const int old_optind = optind;
+    const int old_optopt = optopt;
+    char * const old_optarg = optarg;
```
Comment (beacer) [inline src/main.c]: `const char *old_optarg = ...`

File: src/main.c [hunk 6]
```diff
@@ -73,6 +74,62 @@ static int set_all_thread_affinity(void)
...
+        }
+    }
+
+    if (optind > 0)
+        argv[optind-1] = prgname;
+
+    ret = optind - 1;
+
+    /* restore getopt lib */
+    optind = old_optind;
```
Comment (beacer) [inline src/main.c]: If some argument are consumed, is old index still correct ?

File: src/main.c [hunk 7]
```diff
@@ -84,9 +141,26 @@ int main(int argc, char *argv[])
...
+        exit(EXIT_FAILURE);
+    }
+    argc -= err, argv += err;
+
     /* check if dpvs is running and remove zombie pidfile */
     if (dpvs_running(DPVS_PIDFILE)) {
-        fprintf(stderr, "dpvs is already running\n");
+        fprintf(stderr, 
+            "dpvs is already running\ndpvs version: %s, build on %s\n", 
+            DPVS_VERSION,
```
Comment (beacer) [inline src/main.c]: The question I've raised still not get resolved, if print version here, user may get wrong version for old dpvs which is already running. Should not print version here I think.

File: src/main.c (hunk 2 above)
Comment (kldeng) [reply src/main.c]: Done.

File: src/main.c (hunk 5 above)
Comment (kldeng) [reply src/main.c]: char *const is a const pointer to char while const char * is a pointer to const char. We need the pointer not the value to be unchanged. We can also see it at eal_parse_args.

File: src/main.c (hunk 6 above)
Comment (kldeng) [reply src/main.c]: Yes. old_optind won't change after that. And all eal arguments parse function will reset optind to 1 as double assurance.

**Pull request merged**